src/System.Net.WebHeaderCollection/src/System/Net/HeaderInfoTable.cs

```diff
@@ -10,64 +10,137 @@ namespace System.Net
     internal class HeaderInfoTable
     {
         private static Hashtable HeaderHashTable;
-        private static HeaderInfo UnknownHeaderInfo = new HeaderInfo(string.Empty, false, false, false);
+        private static HeaderInfo UnknownHeaderInfo = new HeaderInfo(string.Empty, false, false, false, SingleParser);
+        private static HeaderParser SingleParser = new HeaderParser(ParseSingleValue);
+        private static HeaderParser MultiParser = new HeaderParser(ParseMultiValue);
+        private static HeaderParser SetCookieParser = new HeaderParser(ParseSetCookieValue);
+        
+        private static string[] ParseSingleValue(string value)
+        {
+            return new string[1]{value};
+        }
+
+        private static string[] ParseMultiValue(string value)
+        {
+            return ParseValueHelper(value, false);
+        }
+
+        private static string[] ParseSetCookieValue(string value)
+        {
+            return ParseValueHelper(value, true);
+        }
+
+        private static string[] ParseValueHelper(string value, bool isSetCookie)
+        {
+            StringCollection tempStringCollection = new StringCollection();
+
+            bool inquote = false;
+            int chIndex = 0;
+            char[] vp = new char[value.Length];
```

Comment:
As an aside, in a separate PR at some point, this could be cleaned up to avoid unnecessary allocations.